lib/services/db.js

```diff
@@ -215,6 +216,91 @@ function formatBatchOperations(ops) {
   }).join('\n');
 }
 
+/**
+ *
+ * @param {object} hookOps
+ * @param {Array} hookOps.put
+ * @param {Array} hookOps.del
+ * @param {object} op
+ * @returns {object}
+ */
+function addToHookOps(hookOps, op) {
+  if (hookOps[op.type]) {
+    hookOps[op.type].push(_.pick(op, 'type', 'key', 'value'));
+  }
+  return hookOps;
+}
+
+/**
+ *
+ * @param {string} type
+ * @param {string} key
+ * @param {*} value
+ * @returns {[{}]}
+ */
+function singleOpToBatch(type, key, value) {
+  return [{type: type, key: key, value: value}];
+}
+
+/**
+ * reduce database operations to key-value arrays sorted by operation type
+ * defines what the plugin hooks expect to receive
+ * @param {string} method
+ * @param {Array} args    args from put, del, or batch
+ * @returns {{put: [{key: value}], del: [{key: value}]}}
+ */
+function dbOpsToHookOps(method, args) {
+  var hookOps = {
+      put: [],
+      del: []
+    },
+    batchOps = method === 'batch' ? args[0] : singleOpToBatch(method, args[0], args[1]);
+
+  return _.reduce(batchOps, addToHookOps, hookOps);
+}
+
+/**
+ *
+ * @param {string} method
+ * @param {Array} args
+ */
+function triggerPlugins(method, args) {
+  const hookOps = dbOpsToHookOps(method, args),
+    // assumes page publish batches always have page put as the last op
+    lastPutKey = _.get(_.last(hookOps.put), 'key'),
+    isPublishedPageBatch = method === 'batch' && hookOps.put.length > 1 &&  _.includes(lastPutKey, '/pages/') && _.endsWith(lastPutKey, publishedVersionSuffix);
+
+  if (hookOps.put.length) {
+    plugins.executeHook('save', hookOps.put);
+    if (isPublishedPageBatch) {
+      plugins.executeHook('publish', {
```

Comment:
Is there a use case for not wanting to just process a `save` of published data as a `publish` hook? This would give a handler two chances to run on one save of data and I don't know if that's a little too flexible. Thoughts?